File: js/hqc/src/rm.js
/**
 * Reed-Muller code RM(1, 7) for HQC.
 *
 * RM(1, 7) encodes 8 bits (1 byte) into 128 bits.
 * The encoding uses the generator matrix of RM(1, 7).
 *
 * For HQC, the RM codeword is duplicated (multiplicity times)
 * to form an n2-bit codeword for additional error correction.
 */

const RM_BASE_LEN = 128;

/**
 * Encode a single byte into a 128-bit RM(1,7) codeword.
 * Returns [lo32_0, lo32_1, lo32_2, lo32_3] as four 32-bit words (128 bits).
 * @param {number} msg - byte to encode (0-255)
 * @returns {Uint32Array} - 4 words representing 128 bits
 */
export function rmEncodeBase(msg) {
  // Generator matrix rows for RM(1,7) as 128-bit patterns:
  // We work with 4 x 32-bit words.
  //
  // Row 0 (constant): all-ones
  // Row 1: 0xAAAAAAAA repeated
  // Row 2: 0xCCCCCCCC repeated
  // Row 3: 0xF0F0F0F0 repeated
  // Row 4: 0xFF00FF00 repeated
  // Row 5: 0xFFFF0000 repeated
  // Row 6: (0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF)
  // Row 7: (0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF)

  const out = new Uint32Array(4);

  const expand = (bit) => -((msg >>> bit) & 1); // all-0 or all-1

  // Bit 0: constant row (all-ones if set)
  const m0 = expand(0);
  out[0] ^= m0; out[1] ^= m0; out[2] ^= m0; out[3] ^= m0;

  // Bit 1: 0xAAAAAAAA
  const m1 = expand(1) & 0xAAAAAAAA;
  out[0] ^= m1; out[1] ^= m1; out[2] ^= m1; out[3] ^= m1;

  // Bit 2: 0xCCCCCCCC
  const m2 = expand(2) & 0xCCCCCCCC;
  out[0] ^= m2; out[1] ^= m2; out[2] ^= m2; out[3] ^= m2;

  // Bit 3: 0xF0F0F0F0
  const m3 = expand(3) & 0xF0F0F0F0;
  out[0] ^= m3; out[1] ^= m3; out[2] ^= m3; out[3] ^= m3;

  // Bit 4: 0xFF00FF00
  const m4 = expand(4) & 0xFF00FF00;
  out[0] ^= m4; out[1] ^= m4; out[2] ^= m4; out[3] ^= m4;

  // Bit 5: 0xFFFF0000
  const m5 = expand(5) & 0xFFFF0000;
  out[0] ^= m5; out[1] ^= m5; out[2] ^= m5; out[3] ^= m5;

  // Bit 6: (0x00000000, 0xFFFFFFFF, 0x00000000, 0xFFFFFFFF)
  const m6 = expand(6);
  out[1] ^= m6; out[3] ^= m6;

  // Bit 7: (0x00000000, 0x00000000, 0xFFFFFFFF, 0xFFFFFFFF)
  const m7 = expand(7);
  out[2] ^= m7; out[3] ^= m7;

  return out;
}

/**
 * Encode a byte into dst starting at bitOffset, with repetition.
 * @param {Uint32Array} dst
 * @param {number} msg
 * @param {number} bitOffset
 * @param {number} multiplicity
 */
export function rmEncodeInto(dst, msg, bitOffset, multiplicity) {
  const base = rmEncodeBase(msg);

  let bitPos = bitOffset;
  for (let rep = 0; rep < multiplicity; rep++) {
    for (let w = 0; w < 4; w++) {
      const word = base[w];
      const dstWord = bitPos >>> 5;
      const dstBit = bitPos & 31;

      if (dstBit === 0 && dstWord < dst.length) {
        // Aligned write
        dst[dstWord] ^= word;
        bitPos += 32;
      } else {
        // Unaligned: write bit by bit
        for (let bit = 0; bit < 32; bit++) {
          if (word & (1 << bit)) {
            const idx = bitPos >>> 5;
            const off = bitPos & 31;
            if (idx < dst.length) {
              dst[idx] ^= 1 << off;
            }
          }
          bitPos++;
        }
      }
    }
  }
}

/**
 * Decode an n2-bit received codeword (with duplicated RM(1,7)) to a single byte
 * using the Walsh-Hadamard transform.
 * @param {Uint32Array} src
 * @param {number} n2
 * @param {number} multiplicity
 * @returns {number} decoded byte
 */
export function rmDecode(src, n2, multiplicity) {
  // Step 1: accumulate all copies into signed sums of 128 entries
  const sums = new Int32Array(RM_BASE_LEN);

  let bitPos = 0;
  for (let rep = 0; rep < multiplicity; rep++) {
    for (let i = 0; i < RM_BASE_LEN; i++) {
      const wordIdx = bitPos >>> 5;
      const bitIdx = bitPos & 31;
      let bit = 0;
      if (wordIdx < src.length) {
        bit = (src[wordIdx] >>> bitIdx) & 1;
      }
      // Convert 0/1 to +1/-1 (0 -> +1, 1 -> -1)
      sums[i] += 1 - 2 * bit;
      bitPos++;
    }
  }

  // Step 2: Fast Walsh-Hadamard Transform (in-place, 7 passes)
  for (let pass = 0; pass < 7; pass++) {
    const step = 1 << pass;
    for (let i = 0; i < RM_BASE_LEN; i += 2 * step) {
      for (let j = i; j < i + step; j++) {
        const a = sums[j];
        const b = sums[j + step];
        sums[j] = a + b;
        sums[j + step] = a - b;
      }
    }
  }

  // Step 3: Find position with maximum absolute value
  let maxAbs = 0;
  let maxPos = 0;
  let sign = 1;

  for (let i = 0; i < RM_BASE_LEN; i++) {
    const v = sums[i];
    const abs = v < 0 ? -v : v;
    if (abs > maxAbs) {
      maxAbs = abs;
      maxPos = i;
      sign = v > 0 ? 1 : -1;
    }
  }

  // Step 4: Recover message byte
  let msg = (maxPos << 1) & 0xFF;
  if (sign < 0) msg |= 1;
  return msg;
}

File: rust/hqc/src/rm.rs
/// Reed-Muller code RM(1, 7) for HQC.
///
/// The first-order Reed-Muller code RM(1, 7) encodes 8 bits (1 byte)
/// into 128 bits (16 bytes). The encoding uses the generator matrix
/// of RM(1, 7), which consists of the all-ones word and the 7 rows
/// of the Walsh-Hadamard basis vectors.
///
/// For HQC, the RM codeword is then duplicated (`multiplicity` times)
/// to form an n2-bit codeword for additional error correction.

/// Base Reed-Muller codeword length = 2^7 = 128 bits.
const RM_BASE_LEN: usize = 128;

/// Encodes a single byte (8 bits) into a 128-bit RM(1,7) codeword.
/// Returns `[lo, hi]` representing 128 bits.
pub fn rm_encode_base(msg: u8) -> [u64; 2] {
    let expand = |bit: u32| -> u64 {
        -(((msg >> bit) & 1) as i64) as u64
    };

    let mut lo = 0u64;
    let mut hi = 0u64;

    // Bit 0: constant row (all-ones if set)
    lo ^= expand(0);
    hi ^= expand(0);

    // Bit 1: pattern 0xAAAAAAAAAAAAAAAA
    lo ^= expand(1) & 0xAAAAAAAAAAAAAAAA;
    hi ^= expand(1) & 0xAAAAAAAAAAAAAAAA;

    // Bit 2: pattern 0xCCCCCCCCCCCCCCCC
    lo ^= expand(2) & 0xCCCCCCCCCCCCCCCC;
    hi ^= expand(2) & 0xCCCCCCCCCCCCCCCC;

    // Bit 3: pattern 0xF0F0F0F0F0F0F0F0
    lo ^= expand(3) & 0xF0F0F0F0F0F0F0F0;
    hi ^= expand(3) & 0xF0F0F0F0F0F0F0F0;

    // Bit 4: pattern 0xFF00FF00FF00FF00
    lo ^= expand(4) & 0xFF00FF00FF00FF00;
    hi ^= expand(4) & 0xFF00FF00FF00FF00;

    // Bit 5: pattern 0xFFFF0000FFFF0000
    lo ^= expand(5) & 0xFFFF0000FFFF0000;
    hi ^= expand(5) & 0xFFFF0000FFFF0000;

    // Bit 6: pattern 0xFFFFFFFF00000000
    lo ^= expand(6) & 0xFFFFFFFF00000000;
    hi ^= expand(6) & 0xFFFFFFFF00000000;

    // Bit 7: (0x0000000000000000, 0xFFFFFFFFFFFFFFFF)
    hi ^= expand(7);

    [lo, hi]
}

/// Encodes a byte into the dst vector starting at `bit_offset`,
/// repeating the base codeword `multiplicity` times.
pub fn rm_encode_into(dst: &mut [u64], msg: u8, bit_offset: usize, multiplicity: usize) {
    let base = rm_encode_base(msg);

    let mut bit_pos = bit_offset;
    for _rep in 0..multiplicity {
        for w in 0..2 {
            let word = base[w];
            let dst_word = bit_pos / 64;
            let dst_bit = bit_pos % 64;

            if dst_bit == 0 && dst_word < dst.len() {
                dst[dst_word] ^= word;
                bit_pos += 64;
            } else {
                for bit in 0..64 {
                    if word & (1u64 << bit) != 0 {
                        let idx = bit_pos / 64;
                        let off = bit_pos % 64;
                        if idx < dst.len() {
                            dst[idx] ^= 1u64 << off;
                        }
                    }
                    bit_pos += 1;
                }
            }
        }
    }
}

/// Decodes an n2-bit received codeword (with duplicated RM(1,7))
/// to a single byte using Walsh-Hadamard transform.
pub fn rm_decode(src: &[u64], n2: usize, multiplicity: usize) -> u8 {
    // Step 1: Accumulate all copies into signed sums
    let mut sums = [0i32; RM_BASE_LEN];

    let mut bit_pos = 0usize;
    for _rep in 0..multiplicity {
        for i in 0..RM_BASE_LEN {
            let word_idx = bit_pos / 64;
            let bit_idx = bit_pos % 64;
            let bit = if word_idx < src.len() {
                ((src[word_idx] >> bit_idx) & 1) as i32
            } else {
                0
            };
            // Convert 0/1 to +1/-1 (0 -> +1, 1 -> -1)
            sums[i] += 1 - 2 * bit;
            bit_pos += 1;
        }
    }
    let _ = n2; // n2 is implicitly multiplicity * 128

    // Step 2: Fast Walsh-Hadamard Transform (in-place, 7 passes)
    for pass in 0..7 {
        let step = 1 << pass;
        let mut i = 0;
        while i < RM_BASE_LEN {
            for j in i..i + step {
                let a = sums[j];
                let b = sums[j + step];
                sums[j] = a + b;
                sums[j + step] = a - b;
            }
            i += 2 * step;
        }
    }

    // Step 3: Find position with maximum absolute value
    let mut max_abs = 0i32;
    let mut max_pos = 0usize;
    let mut sign = 1i32;

    for i in 0..RM_BASE_LEN {
        let v = sums[i];
        let abs = v.abs();
        if abs > max_abs {
            max_abs = abs;
            max_pos = i;
            sign = if v > 0 { 1 } else { -1 };
        }
    }

    // Step 4: Recover message byte
    let mut msg = (max_pos << 1) as u8;
    if sign < 0 {
        msg |= 1;
    }
    msg
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_rm_roundtrip_all_bytes() {
        for m in 0..=255u8 {
            for &mult in &[3, 5] {
                let n2 = mult * 128;
                let n_words = (n2 + 63) / 64;
                let mut cw = vec![0u64; n_words];
                rm_encode_into(&mut cw, m, 0, mult);
                let decoded = rm_decode(&cw, n2, mult);
                assert_eq!(decoded, m, "RM mult={} msg={}: got {}", mult, m, decoded);
            }
        }
    }
}
